pii_recognition/evaluation/character_level_evaluation.py

```diff
@@ -0,0 +1,114 @@
+from dataclasses import asdict
+from typing import Any, Dict, List
+
+from pii_recognition.evaluation.metrics import (
+    compute_label_precision,
+    compute_label_recall,
+)
+from pii_recognition.labels.schema import SpanLabel
+
+
+def encode_labels(
+    text_length: int, span_labels: List[SpanLabel], label_to_int: Dict[Any, int],
+) -> List[int]:
+    """Encode span-based labels into integers.
+
+    Encode a text at character level according to text-span labels as well as a mapping
+    defined by `label_to_int`. Note multi-tagging is not supported. One entity can have
+    only one label tag.
+
+    Args:
+        text_length: length of a text.
+        spans_labels: entities being identified in a text represented by text-spans.
+        label_to_int: a mapping between entity labels and integers.
+
+    Returns:
+        Integer coding of the text.
+    """
+    if 0 in label_to_int.values():
+        raise ValueError(
+            "Value 0 is reserved! If a character does not belong to any classes, it "
+            "will be assigned with 0."
+        )
+    coding = [0] * text_length
+
+    for span in span_labels:
+        s = span.start
+        e = span.end
+        if e > text_length:
+            raise ValueError(
+                f"Span index is out of range: text length is "
+                f"{text_length} but got span index {e}."
+            )
+        label_name = span.entity_type
+        try:
+            label_code = label_to_int[label_name]
+        except KeyError as err:
+            raise Exception(
+                f"Label {str(err)} is not presented in 'label_to_int' mapping."
+            )
+
+        coding[s:e] = [label_code] * (e - s)
```

Comment:
is`span.end` included in the span?  ie. is the span assumed to be a closed interval? (https://mathworld.wolfram.com/Interval.html)

Reply:
`end` is not included.